File: contests/nye18/p4/p4_solution.py
def solve(n, k):
    def is_digit_in_number(number, digit):
        current_digit = 0
        while number != 0:
            current_digit = number % 10
            number //= 10

            if current_digit == digit:
                return True

        return False

    prime = [True]*(n+1)
    p = 2
    while p*p <= n:
        if prime[p]:
            for i in range(p*p,n+1,p):
                prime[i]=False
        p += 1

    prime_sum = 0
    for i in range(2,n+1):
        if prime[i] and is_digit_in_number(i,k):
            prime_sum += i

    return prime_sum

n,k=map(int,input().split())
print(solve(n,k))

File: contests/nye18/p4/p4_solution.cpp
#include <iostream>
#include <string.h>
#include <bitset>
typedef long long ll;

bool digit_in_number(ll digit, ll number)
{
    ll current_digit;
    while (number != 0)
    {
        current_digit = number % 10;
        number /= 10;
                                       
        if (current_digit == digit)
        {
            return true;
        }
    }
    
    return false;
}

#define MAXN 20000001

int main() 
{ 
    ll n;
    int k;
    std::cin >> n >> k;
    
    std::bitset<MAXN> prime;
    prime.set();
    
    for (ll p=2; p*p<=n; p++) 
    { 
        if (prime[p] == true) 
        { 
            for (ll i=p*p; i<=n; i += p)
            {
                prime[i] = false;
            }
        } 
    } 
  
    ll sum = 0;
    for (ll p=2; p<=n; p++)
    {
        if (prime[p] && digit_in_number(k,p))
        {
            sum += p;
        } 
    }
          
    std::cout << sum << "\n";
} 